### ISSUE TITLE
Strange performance behaviour with large arrays

### ISSUE THREAD

**Issue by Twittadrock** [rusty1s/pytorch_scatter#18]
I face a strange performance behaviour in `gen()` function when using "larger" arrays.
Interesstingly this only happens when training with real data (mnistgraph_graclus_spline.py) and I'm not able to reproduce it with a toy example (performance_scatter_add.py). Both files are attached.

From the console log:
> python mnistgraph_graclus_spline.py
Processing...
Done!
Processing...
Done!
forward() >> 0.00 s  data.edge_index.shape torch.Size([2, 1170400])
forward() >> 0.00 s  data.x.shape torch.Size([78400, 1])
forward() >> 0.00 s  data.pos.shape torch.Size([78400, 2])
forward() >> 0.00 s  data.batch tensor([ 0,  0,  0,  ..., 99, 99, 99], device='cuda:0')
forward() >> 0.00 s  data = max_pool(...)
forward() >> 0.02 s  self.conv1(...) torch.Size([78400, 10])
forward() >> 0.02 s  self.normalized_cut_2d(...) torch.Size([78400, 10])
forward() >> 0.03 s  graclus(...) torch.Size([78400, 10])
forward() >> 29.09 s  data.x.shape pooled torch.Size([40697, 10])
forward() >> 0.00 s  data.edge_index.shape torch.Size([2, 1170400])
forward() >> 0.00 s  data.x.shape torch.Size([78400, 1])
forward() >> 0.00 s  data.pos.shape torch.Size([78400, 2])
forward() >> 0.00 s  data.batch tensor([ 0,  0,  0,  ..., 99, 99, 99], device='cuda:0')
forward() >> 0.00 s  data = max_pool(...)
forward() >> 0.00 s  self.conv1(...) torch.Size([78400, 10])
forward() >> 0.00 s  self.normalized_cut_2d(...) torch.Size([78400, 10])
forward() >> 0.01 s  graclus(...) torch.Size([78400, 10])
forward() >> 29.72 s  data.x.shape pooled torch.Size([40712, 10])
forward() >> 0.00 s  data.edge_index.shape torch.Size([2, 1170400])

When profiling it boils down to line 15: in gen.py:
`dim_size = index.max().item() + 1 if dim_size is None else dim_size`

I believe there is something strange with index.max() when the index array is in Cud memory. When I disable Cuda the performance is relatively normal.

Can you reproduce this behavior or have you encountered something similar? 

Thanks in advance!
Matthias

[performance_scatter_add.py.zip](https://github.com/rusty1s/pytorch_scatter/files/2685477/performance_scatter_add.py.zip)

[mnistgraph_graclus_spline.py.zip](https://github.com/rusty1s/pytorch_scatter/files/2685467/mnistgraph_graclus_spline.py.zip)

**Comment by rusty1s** [comment]
Hi,
thanks for this issue. I never used `timeit` by myself, but are you sure this behaves correctly when using PyTorch on GPU? As far as I know, you need to call `torch.cuda.synchronize()` to correctly evaluate durations.

I evaluated the performance by myself using the following script:

```python
import time

import torch
import torch_scatter

N = 1000000
M = 100

x = torch.randn(N, M).cuda()
idx = torch.arange(N).view(-1, 1).expand_as(x).contiguous().cuda()

torch.cuda.synchronize()
t = time.perf_counter()
for i in range(10):
    out = torch.zeros(idx.max().item() + 1, M, device=x.device)
    out.scatter_add_(0, idx, x)
torch.cuda.synchronize()
print(time.perf_counter() - t)

torch.cuda.synchronize()
t = time.perf_counter()
for i in range(10):
    torch_scatter.scatter_add(x, idx, dim=0)
torch.cuda.synchronize()
print(time.perf_counter() - t)
```

which outputs `0.748` seconds for the first run and `0.751` seconds for the second, which shows that both methods run about the same duration.

**Comment by rusty1s** [comment]
Your example code on MNIST looks kinda strange to me, e.g. you set your positions to (1, 1) for all nodes, in which case the Spline Convolution does nothing useful.

However, I think I found the issue, which is in the `max_pool` call when edge attributes are given.
All runtime issues are gone if you set `data.edge_attr=None` before the max pool call:
```python
data.edge_attr = None
data = max_pool(cluster, data, ...)
```
Sadly, I can not do much about it, as this is a problem with the PyTorch Unique function, which is fucking slow :(

**Comment by Twittadrock** [comment]
I'm sorry for the confusion, the MNIST-like-Testset ist just a dummy to reproduce the issue. 

Wow, setting `data.edge_attr = None` really solves the performance/timing issue. 
If I understood correctly, setting `data.edge_attr = None` causes to skip the even larger `torch.unique(.)` in `coalesce,py, line 33`, which implicitly also does a super heavy `scatter_add(.)`?
Maybe we could then change the default behaviour of `coalesce.py` to skip the computation if no `op(.)` is explicitly set..

Anyway thanks for the hint with `torch.cuda.synchronize()`! This cost me a lot of nerves... :))

Currently, I rely heavily on large neighbourhoods and the scatter_add is crucial for performance. Maybe we come up with an optimized version of it or pre-compute max_idx somewhere. 
Do you know how your original unique-cuda-implementation performed in comparison to the pytorch one?

Best wishes,
Matthias

**Comment by rusty1s** [comment]
Hi,

I do not think that the `scatter_{}` ops are a problem, but rather the `unique` calls which are not very efficient on GPU. However, my `unique` implementation is currently in use in [torch_sparse](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/coalesce.py) (when no edge attributes are given) and it runs much, much faster than the PyTorch Unique op. Sadly, it can not handle coalescing of edge attributes, so we default to the PyTorch Unique op in this case. I'm leaving this issue open until I come up with an idea to implement it efficiently.

FYI, you can pre-compute `max_idx` and pass it as the `dim_size` parameter for each `scatter_{}` call. In PyTorch Geometric, we do this all the time:
```python
num_nodes = x.size(0)
row, col = edge_index
out = scatter_add(x[col], row, dim=0, dim_size=num_nodes)
```
